<a href="https://colab.research.google.com/github/jcmm518/03MAIR-Algoritmos-de-optimizacion/blob/main/Seminario_Algoritmos_Juan_Carlos_Marin.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos: **Juan Carlos Marin Mejia**   <br>
Url: https://github.com/.../03MAIR---Algoritmos-de-Optimizacion---2019/tree/master/SEMINARIO<br>

Url Colab: https://colab.research.google.com/drive/1QFa1rxNZFjZz1agwi_DpPy_HeaQew_65?usp=sharing

**Problema:**

> Combinar cifras y operaciones

Descripción del problema:

* Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/)


* Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una cantidad dada. Un ejemplo sería para obtener el 4:4+2-6/3*1 = 4

(*) La respuesta es obligatoria                                      

### Inicio de la solución

### 1. (*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>








### Sin restricciones

Sin tener las restricciones, se podrian repetir los numeros y las operaciones, por lo tanto, se tendría lo siguiente:

Contamos con nueve (9) numeros, independiente de los operaciones, podemos tomar los 9 numeros de formas diferentes, por ejemplo 987654321 , 897645321, 798654213, etc, por lo tanto tenemos que:

        Formas de tomar los numeros = 9 x 9 x 9 x 9 x 9 x 9 x 9 x 9 x 9 = 387.420.489
       

Ahora la pregunta es, ¿cuántas formas de operar dichas combinaciones existen?

Pues en cada combinacion de 9 numeros, podemos operar hasta 8 veces, usando los 4 operadores diferentes, admitiendo cualquier operador tantas veces sea necesario, por ejemplo 1+2*3/4+5-6+7+8-9

        Formas de operar = 4^8 = 65.536 

Combinando, tendríamos lo siguiente:

        Posibilidades = (formas de tomar los numeros) * (formas de operarlos) 
        
        Posibilidades = (387.420.489) * (65.536) = 25.389.989.167.104

**Segun los calculos, sin restricciones, tendriamos 25.389.989.167.104 opciones posibles de operar.**



### 2. ¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones

### Con restricciones

Teniendo en cuenta las restricciones dadas, NO se podrian repetir los numeros y las operaciones, por lo tanto, se tendría lo siguiente:

Contamos con nueve (9) numeros que no se pueden repetir, pero al tener la limitante de poder  utilizar solo 4 operadores que tampoco se pueden repetir, solo podemos contar maximo con tomar 5 digitos de 9, debido a que mas de cinco, no tendriamos operador para trabajarlo, por ejemplo 98721 , 86453, 79813, etc, por lo tanto tenemos que:

        Formas de tomar los números = 9 x 8 x 7 x 6 x 5 = 15.120

Ahora la pregunta es, cuantas formas de operar dichas combinaciones existen ?

Pues en cada combinacion de los 5 numeros, podemos operar hasta 4 veces, usando los 4 operadores diferentes, por ejemplo 4+2-5*3/1

        Formas de operar = 4 x 3 x 2 x 1 = 24

Combinando, tendríamos lo siguiente:

        Posibilidades = (formas de tomar los numeros) * (formas de operarlos)
        
        Posibilidades = (15.120) * (24) = 362.880

**Segun los calculos, con restricciones, tendriamos 362.880 opciones posibles de operar.**


### 3. Modelo para el espacio de soluciones<br>
(*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)


#### Modelo adecuado de estructura de datos

La estructura de datos mas adecuada para dar solucion al problema, es almacenamiento en listas de datos o arrays, debido a que podemos almacenarlos de forma organizada y simplificar los recorridos

Para resolver el algoritmo que entrega solución óptima al problema, se recurre a la técnica: "Divide y Vencerás", se utiliza debido a que resolver el problema completamente por fuerza bruta, resulta en un algoritmo de una enorme complejidad, por lo tanto, se subdivide el problema en dos y una vez que de manera separada se hayan obtenido todas las combinaciones posibles, se vuelven a juntar las 2 partes en que se subdividió el problema,  y así a través de bucles y condicionales contrastar si el resultado de la operación completa, es igual al número inicial ingresado.



### 4. Función objetivo

Según el modelo para el espacio de soluciones<br>
#### (*)¿Cual es la función objetivo?

(*)¿Es un problema de maximización o minimización?

Respuesta

El objetivo del problema, es el de ejecutar un algoritmo que encuentre combinaciones de 5 números enteros entre 1 y 9, que sin estar repetidos y aplicando las 4 operaciones, entregue como resultado un valor igual a un número inicial digitado. Lo anterior, no es una función objetivo, y por lo tanto, no es un problema de maximización y tampoco de minimización, ya que no hay que optimizar ninguna función.


### 5. Diseña un algoritmo para resolver el problema por fuerza bruta

In [ ]:
# impotamos las librerias necesarias
from itertools import * 

# Definimos lista de operadores
signos =['+','-','*','/']

# Definicion de listas de numeros
numeros=['1','2','3','4','5','6','7','8','9']

In [ ]:
#Decorador para tiempos de ejecución
from time import time

#Función para calcular tiempo de ejecución
def calcular_tiempo(f):
    def wrapper(*args, **kwargs):        
        inicio = time()       
        resultado = f(*args, **kwargs)       
        tiempo = time() - inicio
        print("Tiempo de ejecución para algoritmo: "+str(tiempo) + ' Segundos')
        return resultado
    
    return wrapper

In [ ]:
@calcular_tiempo
def fuerza_bruta(num_permuta,sig_permuta,resultado_deseado):
    solucion=0
    ejecuciones = 0
    #Se recorren todos los valores de números y signos posibles
    for i in range(len(num_permuta)): #Recorre toda la lista de permutaciones de números
        for s in range (len(sig_permuta)): #Recorre toda la lista de permutaciones de signos
            resultado=eval(num_permuta[i][0]+sig_permuta[s][0]+num_permuta[i][1]+sig_permuta[s][1]+num_permuta[i][2]+sig_permuta[s][2]+num_permuta[i][3]+sig_permuta[s][3]+num_permuta[i][4])
            ejecuciones = ejecuciones + 1
            if resultado==resultado_deseado:
                solucion=(num_permuta[i][0],sig_permuta[s][0],num_permuta[i][1],sig_permuta[s][1],num_permuta[i][2],sig_permuta[s][2],num_permuta[i][3],sig_permuta[s][3],num_permuta[i][4])
                break
                
    # Se imprimen los resultados
    print("La combinacion encontrada se debe formar en orden de las operacoiones MDSR (Multiplicacion, Division, Suma, Resta) ")
    print("\n")
    print("Combinacion encontrada : " + str(solucion)+ " = " + str(resultado_deseado))
    print("\n")
    print("Numero de iteraciones para encontrar solucion : " + str(ejecuciones) )
    print("\n")
    return solucion,ejecuciones

In [ ]:
# se define el programa principal 

# Calculamos las permutaciones
sig_permutados=list(permutations(signos))
num_permutaciones=list(permutations(numeros,5))

# llamado al programa de fuerza bruta
numero_deseado=18
Ec_solucion,ejec=fuerza_bruta(num_permutaciones,sig_permutados,numero_deseado)

La combinacion encontrada se debe formar en orden de las operacoiones MDSR (Multiplicacion, Division, Suma, Resta) 


Combinacion encontrada : ('9', '+', '8', '*', '7', '/', '4', '-', '5') = 18


Numero de iteraciones para encontrar solucion : 349361


Tiempo de ejecución para algoritmo: 2.621835231781006 Segundos


#### Calcula la complejidad del algoritmo por fuerza bruta

La complejidad esta dada por recorrer dos bucles "for", los cuales recorren las permutaciones de los numeros y las permutaciones de los signos, lo cual implica una complejidad de orden NxN  = O(N^2) adicional el calculo de la funcion eval y el condicional que compara si el resultado se encontro, estas dos son constantes y pueden ser despreciables al momento de calcular la complejidad, por lo tanto, nos quedaremos con una complejidad O(N^2)

### 6. (*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

Para resolver el algoritmo que entrega solución óptima al problema, recurrimos a la técnica: "Divide y Vencerás".
La técnica de 'Divide y Vencerás', se utiliza ya que resolver el problema completamente por fuerza bruta, resulta en un algoritmo de alta complejidad, por lo tanto, se subdivide el problema en dos.
Esta solución es mucho más rápida que la entregada por fuerza bruta, siempre y cuando utilicemos los mismos dígitos arrojados por el primero, pero a medida que lo forzamos a realizarlo con muchos dígitos, el tiempo es más largo.
Adicional, en la solución implementada, mejore todo el manejo de paréntesis para representar no solo las cifras y los operandos si no toda la ecuación.
.

In [ ]:
# Importamos las librerias necesarias
import queue

In [ ]:
# Definimos funcion para ingresar el valor objetivo a calcular y los numeros que deseamos utilizar
def ReadInput () :
    s = input("Ingrese los numero que desea utilizar (Separados por coma): ")
    parts = []
    parts = s.split(',')
    a = []


    for   i in range(len(parts)) :
        a.append(int(parts[i])) #int(parts[i])
        i=i+1

    valueObj = int(input("Ingrese el numero objetivo a calcular: "))
    return a,valueObj

In [ ]:
# Dafinimos funcion para resolver por divide y venceras e imprimirlo
@calcular_tiempo
def SolveAndPrint (numbers,targetValue) :
    targetKey = (targetValue << len(numbers)) + (1 << len(numbers)) - 1
    solvedKeys  = [] # HashSet
    keyToLeftParent  = {}
    keyToRightParent  = {}
    keyToOperator = {}
    q = queue.Queue()


    for i in range(len(numbers)) :
        key = (numbers[i] << len(numbers)) + (1 << i)
        solvedKeys.append(key)
        q.put(key)

    while (not q.empty() and not targetKey in solvedKeys ) :

        curKey = q.get()
    
        curMask = curKey & ((1 << len(numbers)) - 1)
        curValue = curKey >> len(numbers)

        keys = []
        keys = solvedKeys.copy()

        for i in range(len(keys)) :
            mask = keys[i] & ((1 << len(numbers)) - 1)
            value = keys[i] >> len(numbers)

            if ((mask & curMask) == 0) :

                for op in range(6) :
                    opSign = '\0'
                    newValue = 0

                    if op == 0 :
                        newValue = curValue + value
                        opSign = '+'
           
                    if op == 1 :
                        newValue = curValue - value
                        opSign = '-'
           
          
                    if op == 2 :
                        newValue = value - curValue
                        opSign = '-'
           

                    if op == 3 :
                        newValue = curValue * value
                        opSign = '*'

                    if op == 4 :
                        newValue = -1
                        if (value != 0 and curValue % value == 0) :
                            newValue = int(curValue / value)
                            opSign = '/'
              
                    if op == 5 :
                        newValue = -1
                        if (curValue != 0 and value % curValue == 0) :
                            newValue = int(value /  curValue)
                            opSign = '/'
              
                    if (newValue >= 0) :
                        newMask = (curMask | mask)
                        newKey = (newValue << len(numbers)) + newMask

                        if (not newKey in  solvedKeys) :
                            solvedKeys.append(newKey)
                            if (op == 2 or op == 5) :
                                keyToLeftParent[newKey]=keys[i]
                                keyToRightParent[newKey]= curKey
                            else :
                                keyToLeftParent[newKey]= curKey
                                keyToRightParent[newKey] = keys[i]
                            keyToOperator[newKey]= opSign
                            q.put(newKey)

    if (not targetKey in solvedKeys) :
        print("Solucion no encontrada")
    else :
        PrintExpression(keyToLeftParent, keyToRightParent, keyToOperator, targetKey, len(numbers))
        global solucion
        solucion = solucion + " = " + str(targetValue)
        print(solucion)



In [ ]:
# Defino funcion de imprimir resultado
def PrintExpression (keyToLeftParent,keyToRightParent,keyToOperator,key, numbersCount) :
    global solucion
    if (not key in  keyToOperator) :
        solucion = solucion + str(key >> numbersCount)
    else :
        solucion=solucion + "("
        PrintExpression(keyToLeftParent, keyToRightParent, keyToOperator, keyToLeftParent[key], numbersCount)   
        solucion = solucion + str(keyToOperator[key])
        PrintExpression(keyToLeftParent, keyToRightParent, keyToOperator, keyToRightParent[key], numbersCount)
        solucion = solucion + ")"



In [ ]:
# llamado principal
solucion =""
a,valueObj = ReadInput ()
SolveAndPrint(a,valueObj)

Ingrese los numero que desea utilizar (Separados por coma): 9,8,7,4,5
Ingrese el numero objetivo a calcular: 18
(4+(7*(8/(9-5)))) = 18
Tiempo de ejecución para algoritmo: 0.004666805267333984 Segundos


#### (*)Calcula la complejidad del algoritmo 

La complejidad del algoritmo disminuye notablemente, pasando de 2.62 segundos a 0.004 segundos, utlizando las mismas operaciones y las mismas cifras, esto se logra mediante la utilizcion de la tecnica divide y venceras y la recursividad que esta inyecta en el codigo

Segun el analisis de operaciones y calculo de la complejidad, este nuevo algoritmo, incluyendo las optimizaciones para manejo de pilas y manejo de parentesis, es de orden O(log N)

#### Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

Se genera procedimiento aleatorio, para búsqueda de datos de entrada y valor objetivo, estos son generados y organizados para ser procesados.

In [ ]:
# Importamos las librerias necesarias
import random
import queue
# Definimos lista de operadores
signos =['+','-','*','/']

num = []
numdv = []

longitud = len(num)
while longitud < 5 :
    numero=str(random.randint(1, 9))
    if not numero in num :
        num.append(numero)
        numdv.append(int(numero))
    longitud = len(num)

num.sort()
target    = random.randint(1, 75)
print("Informacion del experimento")
print("Numeros a utilizar : " + str(num))
print("Objetivo a lograr  : " + str(target))

Informacion del experimento
Numeros a utilizar : ['1', '2', '6', '7', '9']
Objetivo a lograr  : 67


#### Aplica el algoritmo al juego de datos generado

Aplicamos los datos aleatorio de entrada y generamos una solución.

In [ ]:
# Calculamos las permutaciones
sig_permutados=list(permutations(signos))
num_permutaciones=list(permutations(num,5))
solucion =""

# llamado al programa de fuerza bruta
print("Datos algoritmo de fuerza bruta \n")
numero_deseado=target
ff,fc=fuerza_bruta(num_permutaciones,sig_permutados,numero_deseado)

# con divide y venceras
print("\n")
print("Datos algoritmo divide y venceras \n")
SolveAndPrint (numdv,target)

Datos algoritmo de fuerza bruta 

La combinacion encontrada se debe formar en orden de las operacoiones MDSR (Multiplicacion, Division, Suma, Resta) 


Combinacion encontrada : ('9', '*', '7', '+', '6', '-', '2', '/', '1') = 67


Numero de iteraciones para encontrar solucion : 2552


Tiempo de ejecución para algoritmo: 0.02139139175415039 Segundos


Datos algoritmo divide y venceras 

(6+((1*(7*9))-2)) = 67
Tiempo de ejecución para algoritmo: 0.0051174163818359375 Segundos


#### Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

Las fuentes utilizadas para este trabajo son:

> Algoritmos Heurísticos de la asignatura Algoritmos de Optimización, Dr. Raúl Reyero

> https://eltamiz.com/elcedazo/2011/07/06/resolviendo-cifras-y-letras-ii/

> https://eltamiz.com/elcedazo/2011/06/27/resolviendo-cifras-y-letras-i/

> https://www.geeksforgeeks.org/

> Fundamentos de algoritmia - G. Brassard & P. Bratley

#### Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Considerando la variación en tamaño, el método de fuerza bruta quedaría limitado ya que por la complejidad que maneja llegaría a tiempos de ejecución muy largos, llegando a ser poco eficaz, en el caso del método heuristico por la complejidad lineal es mucho más eficaz, sería interesante jugar con valores y otras restricciones. adicional se podria revisar la opcion de realizar su desarrollo por la tecnica de algoritmos geneticos experimentando con diferentes técnicas de cruce y mutación se pueden conseguir resultados excelentes, especialmente si aumenta el número de dígitos.